### PR TITLE
Rewrite `enarx-pr-assign` as consumable with GraphQL

### PR DESCRIPTION
**Pull request by mbestavros**
Rewrites the event-parsing logic in the `pr-assign` action to use Github's GraphQL API.

### CODE REVIEW

File: enarx-pr-assign [hunk 1]
```diff
@@ -8,30 +8,56 @@ class Override(Exception):
...
+            createdAt
+          }
+        }
+      }
+    }
+  }
+}
+
+    """
+    for e in enarxbot.graphql(QUERY, org=repo.organization.name, repo=repo.name, pr=pr.number)['repository']['pullRequest']['timelineItems']['nodes']:
```
Comment (npmccallum) [inline enarx-pr-assign:51]: This line is too long.

File: enarx-pr-assign [hunk 2]
```diff
@@ -8,30 +8,56 @@ class Override(Exception):
...
+        if e['__typename'] not in ("PullRequestReview", "ReviewRequestedEvent", "ReviewRequestRemovedEvent"):
+            continue
+        yield {
+            "when": e['createdAt'],
+            "what": e['state'] if e['__typename'] == "PullRequestReview" else e['__typename'],
+            "who": e['author']['login'] if e['__typename'] == "PullRequestReview" else e['requestedReviewer']['login']
```
Comment (npmccallum) [inline enarx-pr-assign:57]: This can be done efficiently with a graphql alias: https://graphql.org/learn/queries/#aliases

File: enarx-pr-assign [hunk 3]
```diff
@@ -8,30 +8,56 @@ class Override(Exception):
...
 
-            yield {
-                "when": e.created_at,
-                "what": e.event.upper(),
-                "who": e.requested_reviewer.login
+    QUERY = """
+query($org:String!, $repo:String!, $pr:Int!) {
+  repository(owner: $org, name: $repo) {
+    pullRequest(number: $pr) {
+      timelineItems(first:100) {
```
Comment (npmccallum) [inline enarx-pr-assign:18]: You definitely need to pull all the events. Use depaginization here.

**Comment by mbestavros** [comment]
Latest push adds depaginization and makes the query easier to parse. Depaginization makes it work as expected, so marking it ready for review.

File: enarx-pr-assign [hunk 4]
```diff
@@ -40,14 +75,14 @@ def get_responsible(pr):
...
     participants = {x["who"] for x in events} - {pr.user.login}
     segregated = {u: [x["what"] for x in events if x["who"] == u] for u in participants}
 
     for (who, whats) in segregated.items():
         unrequested = False
         for what in reversed(whats):
-            if what == "REVIEW_REQUESTED" and not unrequested:
+            if what == "ReviewRequestedEvent" and not unrequested:
                 yield who
             elif what in ("DISMISSED", "PENDING"):
```
Comment (haraldh) [inline enarx-pr-assign:87]: Did this not change to lowercase?

File: enarx-pr-assign [hunk 5]
```diff
@@ -56,17 +91,18 @@ def get_responsible(pr):
             elif what == "APPROVED":
```
Comment (haraldh) [inline enarx-pr-assign:91]: and this?

File: enarx-pr-assign (hunk 4 above)
Comment (mbestavros) [reply enarx-pr-assign:87]: Yes, this is correct. The previous version yielded everything in all caps; this version yields data directly from Github. Pull request events (like `ReviewRequestedEvent`) are first-letter capitalized, as changed above, but PR review states (like `DISMISSED`, `PENDING`, `APPROVED`) are all caps.

File: enarx-pr-assign (hunk 5 above)
Comment (mbestavros) [reply enarx-pr-assign:91]: See above.

**Review by haraldh — APPROVED** [review]

**Review by haraldh — APPROVED** [review]

**Review by haraldh — APPROVED** [review]

**Review by haraldh — APPROVED** [review]

**Review by npmccallum — CHANGES_REQUESTED** [review]
1. Use `yield` and `yield from` in `iterusers()`.
2. Remove old file.
3. Add blank line before line 187.
4. Add to yml.

**Review by npmccallum — APPROVED** [review]

**Review by haraldh — APPROVED** [review]

**Pull request merged**